src/main/java/com/limechain/sync/warpsync/SyncedState.java

```diff
@@ -165,4 +197,101 @@ public Hash256 loadStateRoot() {
         Object storedRootState = repository.find(DBConstants.STATE_TRIE_ROOT_HASH).orElse(null);
         return storedRootState == null ? null : Hash256.from(storedRootState.toString());
     }
+
+    /**
+     * Sync our state according to a neighbour message we receive.
+     * Try to update our set data (id and authorities) if neighbour has a greater set id than us.
+     * Synchronized to avoid race condition between checking and updating set id
+     * @param neighbourMessage received neighbour message
+     * @param peerId sender of message
+     */
+    public synchronized void syncNeighbourMessage(NeighbourMessage neighbourMessage, PeerId peerId) {
+        if (warpSyncFinished && neighbourMessage.getSetId().compareTo(setId) > 0) {
+            updateSetData(neighbourMessage.getLastFinalizedBlock().add(BigInteger.ONE), peerId);
+        }
+    }
+
+    private void updateSetData(BigInteger setChangeBlock, PeerId peerId) {
+        BlockResponse response = network.syncBlock(peerId, setChangeBlock);
+        BlockData block = response.getBlocksList().get(0);
+
+        Justification justification = new JustificationReader().read(
+                new ScaleCodecReader(block.getJustification().toByteArray()));
+        boolean verified = justification != null
+                && JustificationVerifier.verify(justification.precommits, justification.round);
+        if (verified) {
+            BlockHeader header = new BlockHeaderReader().read(new ScaleCodecReader(block.getHeader().toByteArray()));
+            handleAuthorityChanges(header.getDigest(), setChangeBlock);
+            BlockHeader blockHeader = new BlockHeaderReader().read(
+                    new ScaleCodecReader(block.getHeader().toByteArray()));
+            this.lastFinalizedBlockNumber = blockHeader.getBlockNumber();
+            this.lastFinalizedBlockHash = new Hash256(blockHeader.getHash());
+            handleScheduledEvents();
+        }
+    }
+
+    public void handleScheduledEvents() {
+        Pair<BigInteger, Authority[]> data = scheduledAuthorityChanges.peek();
+        boolean updated = false;
+        while (data != null) {
+            if (data.getValue0().compareTo(this.getLastFinalizedBlockNumber()) < 1) {
+                authoritySet = data.getValue1();
+                setId = setId.add(BigInteger.ONE);
+                latestRound = BigInteger.ONE;
+                scheduledAuthorityChanges.poll();
+                updated = true;
+            } else break;
+            data = scheduledAuthorityChanges.peek();
+        }
+        if (warpSyncFinished && updated) {
+            new Thread(() -> network.sendNeighbourMessages()).start();
+        }
```

Comment:
```suggestion
            new Thread(() -> network.sendNeighbourMessages()).start();
            persistState();
        }
```

It would be nice idea to persist state after authority set update